### PR TITLE
Update chain for on chain architecture - Closes #5609

### LINKED ISSUE #5609: Update lisk-chain to support the on-chain architecture
### Description

Other than cleaning up the interface, main changes are:

*   Move BaseTransaction implementation into Transaction class
    *   Export base transaction schema from library
    *   Update base transaction schema for “moduleType” and “assetType” and add validation
*   Move isDPoSProtocolComplient logic into isValidSeedReveal
    *   verifyBlock should include the logic to check the reward
*   Move genesis block common block header validation from lisk-genesis into validateGenesisBlock
*   Move dpos.verifyBlockForger into verifyBlock using set validators and slot
*   Move Base transaction validate logic to transaction.validate
*   Add readGenesisBlockJSON(accoutSchemas: AccountSchema[]) function and export from index.ts
*   Add getAccountSchema function to return complete account schema
*   Add getGenesisBlockSchema function to return complete genesis block schema
*   Add function to manage validators in consensus state
    *   CONSENSUS_VALIDATORS = “consensus:validators”
    *   add function setValidators and getValidators
    *   Export CONSENSUS_VALIDATORS from lisk-chain
*   Update statestore account “get”, “getOrDefault” to to take generics where T is unknown by default otherwise T should corresponds to the account schema
    *   ex) stateStore.get&lt;{ token: { balance: bigint }>(address)
*   Add setValidators and getValidators
    *   setValidator should calculate minActiveHeight from the previous state
    *   setValidator should cache numberOfValidators
*   Update applyGenesisBlock to store all the accounts
    *   Validate genesis block properties
    *   Insert init validators
    * Insert genesis accounts
*   Move BlockV2Schema and type definition as BlockHeader from block processor V2
*   Move V2 block validation in validateBlock

```js
interface ChainConstructor {
	db: KVStore;
	genesisBlock: GenesisBlock;
	accounts: { moduleName: string; schema: AccountSchema }[];
	networkIdentifier: Buffer;
	blockTime: number;
	maxPayloadLength: number;
	rewardDistance: number;
	rewardOffset: number;
	rewardMilestones: ReadonlyArray<string>;
	stateBlockSize?: number;
	minBlockHeaderCache?: number;
	maxBlockHeaderCache?: number;
}
​
interface Chain {
	constructor: (input: ChainConstructor) => void;
	lastBlock: Block;
	genesisBlock: GenesisBlock;
	blockAssetSchema: { [version: number]: Schema };
	numberOfValidators: number;
	// Below 2 schema depends on dynamic account schema
	getAccountSchema: () => Schema;
	getGenesisBlockSchema: () => Schema;
	getReward: (height: number) => bigint;
	// Sets last block, set number of validators, setup cache etc
	init: () => Promise<void>;
	// Returns new state store for processing
	newStateStore: (skipLastHeights: number) => Promise<StateStore>;
	// return true if exist, throws if it doesn't match
	genesisBlockExist: (genesisBlock: GenesisBlock) => Promise<boolean>;
	// Validate common headers
        isValidSeedReveal: (block: Block): bool;
	validateGenesisBlockHeader: (block: GenesisBlock) => void;
	applyGenesisBlock: (block: GenesisBlock, stateStore: StateStore) => void;
	// Validate schema, signature, reward (including seed reveal), transaction root and payload size
	validateBlockHeader: (block: Block) => void;
	// verify previous id, block slot, and generator
	verifyBlockHeader: (block: Block, stateStore: StateStore) => void;
	saveBlock: (block: Block) => Promise<void>;
	deleteLastBlock: (block: Block) => Promise<void>;
        getValidator: (timestamp) => { address: Buffer; weight: 0 | 1; minActiveHeight: number };
	getValidators: () => { address: Buffer; weight: 0 | 1; minActiveHeight: number }[];
	// Invalid if length is not equal number of validators, it should set minActiveHeight
	setValidators: (validators: { address: Buffer; weight: 0 | 1 }[], stateStore: StateStore) => void;
}
```

### Motivation

Lisk-chain library includes some of the state change functionality, but it should be removed.

Also, there are dependencies to lisk-transactions which will be removed.

Lisk-chain handles basic levels of consensus, slots logic and selection based on the validator given.

### Acceptance criteria

*   Remove dependency to the lisk-transactions
*   Update the interface and logic according to the definition above
*   All tests are updated

### Related Issues

#5603

### PR DESCRIPTION
**Pull request by shuse2**
### What was the problem?

This PR resolves #5609 

### How was it solved?

- Transaction is moved into chain.
- Basic genesis block is validate / apply is added
- Get/Set validator is added
- Default account schema is updated
- Add seed reveal validation

### How was it tested?

- Update all existing tests
- Migrate tests for genesis block validation, transaction, seed reveal
- Add test for set/get validators
- Check chain library tests description and changed implementation

### CODE REVIEW

File: elements/lisk-chain/src/chain.ts [hunk 1]
```diff
@@ -376,78 +385,58 @@ export class Chain {
...
+		);
+		const [lastBlock] = stateStore.chain.lastBlockHeaders;
+		const nextValidatorSet = [];
+		for (const nextValidator of validators) {
+			const previousInfo = previousValidators.find(pv => pv.address.equals(nextValidator.address));
+			nextValidatorSet.push({
+				...nextValidator,
+				minActiveHeight:
+					// currently processed block is still in process, so the next block height is +2
+					previousInfo !== undefined ? previousInfo.minActiveHeight : lastBlock.height + 2,
```
Comment (ManuGowda) [inline elements/lisk-chain/src/chain.ts:435]: dint quite get this +2 logic, could you detail it out a bit more?

Comment (shuse2) [reply elements/lisk-chain/src/chain.ts:435]: `currently processed block is still in process, so the next block height is +2`
meaning when this function is called `| last block | currently processed block | next block|` and what we want is the height of the next block

File: elements/lisk-chain/src/chain.ts [hunk 2]
```diff
@@ -92,166 +85,156 @@ export class Chain {
...
-			registeredTransactions,
-			accountSchema,
+			registeredBlockHeaders: this._blockAssetSchema,
+			accountSchema: this._accountSchema,
 			minBlockHeaderCache,
 			maxBlockHeaderCache,
 		});
 
-		this._lastBlock = genesisBlock;
+		this._lastBlock = (genesisBlock as unknown) as Block;
```
[the remark this answers is not in the document]
Comment (shuse2) [reply elements/lisk-chain/src/chain.ts:159]: type of lastBlock is Block not genesis block, that's why it needs to be casted... 😢

[the remark this answers is not in the document]
Comment (ManuGowda) [reply elements/lisk-chain/src/chain.ts:159]: mmm 👍

File: elements/lisk-chain/src/transaction.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { codec } from '@liskhq/lisk-codec';
+import { hash, getAddressFromPublicKey } from '@liskhq/lisk-cryptography';
+import { validator, LiskValidationError } from '@liskhq/lisk-validator';
+
+interface TransactionInput {
```
[the remark this answers is not in the document]
Comment (shuse2) [reply elements/lisk-chain/src/transaction.ts:19]: I think it shouldnt be required.
As long as the input is correct, it shouldnt be required to be typed outside

**Pull request merged**